devito/interfaces.py

```diff
@@ -1,22 +1,110 @@
+                 be given as keyword argument.
+    :param shape: Shape of the underlying spatial data. Must be given
+                  as keyword argument.
+
+    This class implements the symbolic behaviour of Devito's data
+    objects by inheriting from and mimicking the behaviour of :class
+    sympy.Function:. In order to maintain meta information across the
+    numerous re-instantiation SymPy performs during symbolic
+    manipulation, we inject the symbol name as the class name and
+    cache all created objects on that name. This entails that data
+    object should implement `__init__` in the following format:
+
+    def __init__(self, *args, **kwargs):
+        if self._cached():
+            SymbolicData.__init__(self)
+            return
+        else:
+            ... # Initialise object properties from kwargs
+
+            self._cache_put(self)
+
+    Note: The parameters :param name: and :param shape: must always be
+    present and given as keyword arguments, since SymPy uses `*args`
+    to (re-)create the dimension arguments of the symbolic function.
+    """
+
+    def __new__(cls, *args, **kwargs):
+        if cls not in _SymbolCache:
+            name = kwargs.get('name')
+            shape = kwargs.get('shape')
+            if len(args) < 1:
+                args = cls.indices(shape)
+            cls.__name__ = name  # Inject the given variable name
+        return Function.__new__(cls, *args)
+
+    def __init__(self):
+        """Initialise from a cached instance by shallow copying __dict__"""
+        original = _SymbolCache[self.__class__]
+        self.__dict__ = original.__dict__.copy()
+
+    @classmethod
+    def _cached(cls):
+        """Test if current class is already in the symbol cache"""
+        return cls in _SymbolCache
+
+    @classmethod
+    def _cache_put(cls, obj):
+        """Store given object instance in symbol cache"""
+        _SymbolCache[cls] = obj
+
+    @classmethod
+    def indices(cls, shape):
+        """Abstract class method to determine the default dimension indices."""
+        raise NotImplementedError('Abstract class `SymbolicData` does not have default indices')
+
+
+class DenseData(SymbolicData):
+    """Data object for spatially varying data that acts as a Function symbol
+
+    :param name: Name of the resulting :class sympy.Function: symbol
+    :param shape: Shape of the spatial data grid
+    :param dtype: Data type of the buffered data
+
+    Note: :class DenseData: objects are assumed to be constant in time and
+    therefore do not support time derivatives. Use :class TimeData: for
+    time-varying griad data.
+    """
+    def __init__(self, *args, **kwargs):
+        if self._cached():
+            # Initialise instance from symbol cache
+            SymbolicData.__init__(self)
+            return
+        else:
+            self.name = kwargs.get('name')
+            self.shape = kwargs.get('shape')
```

Comment:
Is the Symbol class not immutable? Is that why you can change the shape after the object has been created?
I ask because such a line would not work if the superclass was IndexedBase.